### PR TITLE
Support cuDNN FP16 batch normalization

### PR DESCRIPTION
**Pull request by gwtnb**
`cudnnDeriveBNTensorDescriptor` returns FP32 when `xDesc` is FP16, so cuDNN assumes FP32 `gamma`, `beta`, etc. in FP16 BN. `x` and `gamma` do not necessarily have the same dtype, so we have to compare `gamma.dtype` with `dtype_param` before type conversion.

This change is especially important for `batch_normalization_backward` because we need FP32 `ggamma` for FP32 `gamma`.

### CODE REVIEW

**Comment by gwtnb** [comment]
Jenkins, test this please

**Comment by okuta** [comment]
LGTM!
